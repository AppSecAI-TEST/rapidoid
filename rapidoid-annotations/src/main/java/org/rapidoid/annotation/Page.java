package org.rapidoid.annotation;

/*
 * #%L
 * rapidoid-annotations
 * %%
 * Copyright (C) 2014 - 2015 Nikolche Mihajlovski and contributors
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;

@Target({ METHOD })
@Retention(RUNTIME)
@Authors("Nikolche Mihajlovski")
@Since("4.1.0")
public @interface Page {

	String value() default "";

	boolean raw() default false;

	boolean navbar() default true;

	boolean search() default false;

	boolean profile() default true;

	boolean login() default true;

	String title() default "";

}
