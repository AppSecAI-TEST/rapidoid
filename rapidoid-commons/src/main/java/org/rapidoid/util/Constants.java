package org.rapidoid.util;

/*
 * #%L
 * rapidoid-commons
 * %%
 * Copyright (C) 2014 Nikolche Mihajlovski
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

public interface Constants {

	byte SPACE = ' ';

	byte CR = 13;

	byte LF = 10;

	byte[] CR_LF_CR_LF = { CR, LF, CR, LF };

	byte[] CR_LF = { CR, LF };

	byte[] LF_LF = { LF, LF };

	byte[] SPACE_ = { SPACE };

	byte[] CR_ = { CR };

	byte[] LF_ = { LF };

	byte ASTERISK = '?';

	byte EQ = '=';

	byte AMP = '&';

	byte COL = ':';

}
