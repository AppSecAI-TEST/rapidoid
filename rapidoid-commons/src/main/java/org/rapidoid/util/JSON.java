package org.rapidoid.util;

import java.io.OutputStream;

import com.fasterxml.jackson.databind.ObjectMapper;

/*
 * #%L
 * rapidoid-commons
 * %%
 * Copyright (C) 2014 Nikolche Mihajlovski
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

public class JSON {

	private static final ObjectMapper MAPPER = new ObjectMapper();

	public static String stringify(Object value) {
		try {
			return MAPPER.writeValueAsString(value);
		} catch (Exception e) {
			throw U.rte(e);
		}
	}

	public static void stringify(Object value, OutputStream out) {
		try {
			MAPPER.writeValue(out, value);
		} catch (Exception e) {
			throw U.rte(e);
		}
	}

	public static <T> T parse(String json, Class<T> valueType) {
		try {
			return MAPPER.readValue(json, valueType);
		} catch (Exception e) {
			throw U.rte(e);
		}
	}

}
