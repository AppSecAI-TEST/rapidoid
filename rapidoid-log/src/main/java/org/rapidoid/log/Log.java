package org.rapidoid.log;

/*
 * #%L
 * rapidoid-u
 * %%
 * Copyright (C) 2014 - 2015 Nikolche Mihajlovski
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;

public class Log {

	public static final LogLevel TRACE = LogLevel.TRACE;
	public static final LogLevel DEBUG = LogLevel.DEBUG;
	public static final LogLevel AUDIT = LogLevel.AUDIT;
	public static final LogLevel INFO = LogLevel.INFO;
	public static final LogLevel WARN = LogLevel.WARN;
	public static final LogLevel ERROR = LogLevel.ERROR;
	public static final LogLevel SEVERE = LogLevel.SEVERE;

	protected static LogLevel LOG_LEVEL = INFO;

	private static Appendable LOG_OUTPUT = System.out;

	private Log() {
	}

	public static synchronized void setLogLevel(LogLevel logLevel) {
		LOG_LEVEL = logLevel;
	}

	public static synchronized LogLevel getLogLevel() {
		return LOG_LEVEL;
	}

	private static String getCallingClass() {
		StackTraceElement[] trace = Thread.currentThread().getStackTrace();

		for (int i = 2; i < trace.length; i++) {
			String cls = trace[i].getClassName();
			if (!cls.equals(Log.class.getCanonicalName())) {
				return cls;
			}
		}

		return Log.class.getCanonicalName();
	}

	private static void log(Appendable out, LogLevel level, String msg, String key1, Object value1, String key2,
			Object value2, String key3, Object value3, int paramsN) {
		if (level.ordinal() >= LOG_LEVEL.ordinal()) {
			try {
				synchronized (out) {
					out.append(level.name());
					out.append(" | ");
					out.append(Thread.currentThread().getName());
					out.append(" | ");
					out.append(getCallingClass());
					out.append(" | ");
					out.append(msg);

					switch (paramsN) {
					case 0:
						break;

					case 1:
						printKeyValue(out, key1, value1);
						break;

					case 2:
						printKeyValue(out, key1, value1);
						printKeyValue(out, key2, value2);
						break;

					case 3:
						printKeyValue(out, key1, value1);
						printKeyValue(out, key2, value2);
						printKeyValue(out, key3, value3);
						break;

					default:
						throw new IllegalStateException();
					}

					out.append((char) 10);
				}
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}
	}

	private static void printKeyValue(Appendable out, String key, Object value) throws IOException {
		out.append(" | ");
		out.append(key);
		out.append("=");
		out.append(pretty(value));

		if (value instanceof Throwable) {
			Throwable err = (Throwable) value;
			ByteArrayOutputStream stream = new ByteArrayOutputStream();
			err.printStackTrace(new PrintStream(stream));
			out.append("\n");
			out.append(stream.toString());
		}
	}

	private static String pretty(Object value) {
		return null;
	}

	public static synchronized void setLogOutput(Appendable logOutput) {
		LOG_OUTPUT = logOutput;
	}

	private static void log(LogLevel level, String msg, String key1, Object value1, String key2, Object value2,
			String key3, Object value3, int paramsN) {
		log(LOG_OUTPUT, level, msg, key1, value1, key2, value2, key3, value3, paramsN);
	}

	public static void trace(String msg) {
		log(TRACE, msg, null, null, null, null, null, null, 0);
	}

	public static void trace(String msg, String key, Object value) {
		log(TRACE, msg, key, value, null, null, null, null, 1);
	}

	public static void trace(String msg, String key1, Object value1, String key2, Object value2) {
		log(TRACE, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void trace(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(TRACE, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void debug(String msg) {
		log(DEBUG, msg, null, null, null, null, null, null, 0);
	}

	public static void debug(String msg, String key, Object value) {
		log(DEBUG, msg, key, value, null, null, null, null, 1);
	}

	public static void debug(String msg, String key1, Object value1, String key2, Object value2) {
		log(DEBUG, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void debug(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(DEBUG, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void audit(String msg) {
		log(AUDIT, msg, null, null, null, null, null, null, 0);
	}

	public static void audit(String msg, String key, Object value) {
		log(AUDIT, msg, key, value, null, null, null, null, 1);
	}

	public static void audit(String msg, String key1, Object value1, String key2, Object value2) {
		log(AUDIT, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void audit(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(AUDIT, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void info(String msg) {
		log(INFO, msg, null, null, null, null, null, null, 0);
	}

	public static void info(String msg, String key, Object value) {
		log(INFO, msg, key, value, null, null, null, null, 1);
	}

	public static void info(String msg, String key1, Object value1, String key2, Object value2) {
		log(INFO, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void info(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(INFO, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void warn(String msg) {
		log(WARN, msg, null, null, null, null, null, null, 0);
	}

	public static void warn(String msg, String key, Object value) {
		log(WARN, msg, key, value, null, null, null, null, 1);
	}

	public static void warn(String msg, String key1, Object value1, String key2, Object value2) {
		log(WARN, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void warn(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(WARN, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void warn(String msg, Throwable error) {
		warn(msg, "error", error);
	}

	public static void error(String msg) {
		log(ERROR, msg, null, null, null, null, null, null, 0);
	}

	public static void error(String msg, String key, Object value) {
		log(ERROR, msg, key, value, null, null, null, null, 1);
	}

	public static void error(String msg, String key1, Object value1, String key2, Object value2) {
		log(ERROR, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void error(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(ERROR, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void error(String msg, Throwable error) {
		error(msg, "error", error);
	}

	public static void error(Throwable error) {
		error("error occured!", "error", error);
	}

	public static void severe(String msg) {
		log(SEVERE, msg, null, null, null, null, null, null, 0);
	}

	public static void severe(String msg, String key, Object value) {
		log(SEVERE, msg, key, value, null, null, null, null, 1);
	}

	public static void severe(String msg, String key1, Object value1, String key2, Object value2) {
		log(SEVERE, msg, key1, value1, key2, value2, null, null, 2);
	}

	public static void severe(String msg, String key1, Object value1, String key2, Object value2, String key3,
			Object value3) {
		log(SEVERE, msg, key1, value1, key2, value2, key3, value3, 3);
	}

	public static void severe(String msg, Throwable error) {
		severe(msg, "error", error);
	}

}
