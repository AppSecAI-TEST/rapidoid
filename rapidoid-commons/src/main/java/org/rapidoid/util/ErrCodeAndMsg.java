package org.rapidoid.util;

import org.rapidoid.annotation.Authors;
import org.rapidoid.annotation.Since;

@Authors("Nikolche Mihajlovski")
@Since("5.1.0")
public class ErrCodeAndMsg {

	private final int code;
	private final String msg;

	public ErrCodeAndMsg(int code, String msg) {
		this.code = code;
		this.msg = msg;
	}

	public int code() {
		return code;
	}

	public String msg() {
		return msg;
	}
}
