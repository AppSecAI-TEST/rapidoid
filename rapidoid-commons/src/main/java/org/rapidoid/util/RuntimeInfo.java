package org.rapidoid.util;

/*
 * #%L
 * rapidoid-utils
 * %%
 * Copyright (C) 2014 - 2015 Nikolche Mihajlovski and contributors
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.lang.reflect.Method;

import org.rapidoid.annotation.Authors;
import org.rapidoid.annotation.Since;
import org.rapidoid.cls.Cls;

@Authors("Nikolche Mihajlovski")
@Since("2.0.0")
public class RuntimeInfo {

	private static final Method getGarbageCollectorMXBeans;

	static {
		Class<?> manFactory = Cls.getClassIfExists("java.lang.management.ManagementFactory");
		getGarbageCollectorMXBeans = manFactory != null ? Cls.getMethod(manFactory, "getGarbageCollectorMXBeans")
				: null;
	}

	public static String gcInfo() {
		String gcinfo = "";

		if (getGarbageCollectorMXBeans != null) {
			// FIXME resolve rapidoid-beany dependency
			// List<?> gcs = Cls.invokeStatic(getGarbageCollectorMXBeans);
			// for (Object gc : gcs) {
			// gcinfo += " | " + Beany.getPropValue(gc, "name") + " x" + Beany.getPropValue(gc, "collectionCount")
			// + ":" + Beany.getPropValue(gc, "collectionTime") + "ms";
			// }
		}
		return gcinfo;
	}

}
