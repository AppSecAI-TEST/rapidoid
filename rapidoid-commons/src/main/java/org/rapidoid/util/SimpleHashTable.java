package org.rapidoid.util;

import org.rapidoid.RapidoidThing;
import org.rapidoid.annotation.Authors;
import org.rapidoid.annotation.Since;
import org.rapidoid.commons.Rnd;
import org.rapidoid.u.U;

/*
 * #%L
 * rapidoid-commons
 * %%
 * Copyright (C) 2014 - 2017 Nikolche Mihajlovski and contributors
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

@Authors("Nikolche Mihajlovski")
@Since("2.0.0")
public class SimpleHashTable<T> extends RapidoidThing {

	private static final int DEFAULT_BUCKET_SIZE = 8;

	public final SimpleBucket<T>[] buckets;

	public final int factor;

	private final int hashMask;

	private final int xor = Rnd.rnd();

	public SimpleHashTable(int capacity, int bucketSize) {
		this(capacity, bucketSize, true);
	}

	@SuppressWarnings("unchecked")
	public SimpleHashTable(int capacity, int bucketSize, boolean unbounded) {
		U.must(capacity >= 2, "The capacity is too small!");

		if (bucketSize <= 0) bucketSize = DEFAULT_BUCKET_SIZE;

		int factor = Msc.log2(Math.max(capacity / bucketSize, 1));

		int width = (int) Math.pow(2, factor);
		int realCapacity = width * bucketSize;
		U.must(capacity <= realCapacity);

		this.buckets = new SimpleList[width];
		this.factor = factor;
		this.hashMask = Msc.bitMask(factor);

		for (int i = 0; i < buckets.length; i++) {
			buckets[i] = createBucket(bucketSize, unbounded);
		}
	}

	protected SimpleBucket<T> createBucket(int bucketSize, boolean unbounded) {
		U.must(unbounded, "Only unbounded buckets are supported!");
		return new SimpleList<>(bucketSize);
	}

	public void put(long key, T value) {
		bucket(key).add(value);
	}

	public SimpleBucket<T> bucket(long key) {
		int index = index(key);
		return getBucketAt(index);
	}

	public SimpleBucket<T> getBucketAt(int index) {
		SimpleBucket<T> list;

		// after construction, other threads might need some time to see the new references
		while ((list = buckets[index]) == null) {
		}

		return list;
	}

	public int index(long key) {
		return (int) ((key ^ xor) & hashMask);
	}

	public void clear() {
		for (int i = 0; i < buckets.length; i++) {
			clearBucket(i);
		}
	}

	protected void clearBucket(int index) {
		getBucketAt(index).clear();
	}

	public int bucketCount() {
		return buckets.length;
	}
}
