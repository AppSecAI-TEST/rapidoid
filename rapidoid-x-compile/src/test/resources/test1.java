package abc;

/*
 * #%L
 * rapidoid-x-compile
 * %%
 * Copyright (C) 2014 - 2015 Nikolche Mihajlovski and contributors
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */
import static java.lang.annotation.ElementType.*;
import static java.lang.annotation.RetentionPolicy.*;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.util.Date;
import java.util.Map;
import java.util.Map.Entry;

import javax.annotation.Resource;

import org.rapidoid.util.U;
import org.rapidoid.log.Log;

@mixo.Mixin
@Resource
public class Main {

	@Resource
	public String ggg;

	@Resource
	public void setMm(int mm) {
	}

	public static void main(String[] args) {
		Runnable r1 = new Runnable() {
			@Override
			public void run() {
				Log.info("hello1!");
			}
		};

		r1.run();
		Log.info("success!");
	}
}

class Person {

	class Insider {
	}

	String name = "";

	private int age = 0;

	public Person() {
	}

	public Person(String name, int age) {
		this.name = name;
		this.age = age;
	}

}

class PersonService {

	public Person insert(Person person) {
		return person;
	}

	public Object now(Map<String, Object> params) {
		for (Entry<?, ?> entry : params.entrySet()) {
			Object key = entry.getKey();
			Object value = entry.getValue();
			D.print(key, value);
		}
		return new Date();
	}

}
