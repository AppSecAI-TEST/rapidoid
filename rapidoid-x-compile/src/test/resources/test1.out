INFO | main | abc.Main$1 | hello1!
INFO | main | abc.Main | success!
